File: scrimage-core/src/test/scala/com/sksamuel/scrimage/PixelToolsTest.scala
package com.sksamuel.scrimage

import com.sksamuel.scrimage.pixels.PixelTools
import org.scalatest.{FunSuite, Matchers}

class PixelToolsTest extends FunSuite with Matchers {

  val white = 0xFFFFFFFF
  val yellow = 0xFFFFFF00
  val pink = 0xFFFFAFAF

  test("non transparent alpha component") {
    assert(white === java.awt.Color.WHITE.getRGB)
    assert(PixelTools.alpha(white) === 255)
  }

  test("transparent alpha component") {
    assert(PixelTools.alpha(0xDD001122) === 221)
  }

  test("red component") {
    assert(pink === java.awt.Color.PINK.getRGB)
    assert(PixelTools.red(yellow) === 255)
    assert(PixelTools.red(pink) === 255)
  }

  test("blue component") {
    assert(pink === java.awt.Color.PINK.getRGB)
    assert(PixelTools.blue(yellow) === 0)
    assert(PixelTools.blue(pink) === 175)
  }

  test("green component") {
    assert(yellow === java.awt.Color.YELLOW.getRGB)
    assert(PixelTools.green(yellow) === 255)
    assert(PixelTools.green(pink) === 175)
  }

  test("rgb combination") {
    assert(0xFFFF00FF === PixelTools.rgb(255, 0, 255))
    assert(0xFF556677 === PixelTools.rgb(85, 102, 119))
  }

  test("coordinate to offset") {
    assert(160 === PixelTools.coordinateToOffset(10, 3, 50))
    assert(10 === PixelTools.coordinateToOffset(10, 0, 50))
    assert(99 === PixelTools.coordinateToOffset(49, 1, 50))
  }

  test("offset to coordinate") {
    PixelTools.offsetToCoordinate(0, 100) shouldBe new Coordinate(0, 0)
    PixelTools.offsetToCoordinate(100, 100) shouldBe new Coordinate(0, 1)
    PixelTools.offsetToCoordinate(99, 100) shouldBe new Coordinate(99, 0)
    PixelTools.offsetToCoordinate(199, 100) shouldBe new Coordinate(99, 1)
    PixelTools.offsetToCoordinate(101, 100) shouldBe new Coordinate(1, 1)
  }
}

File: scrimage-core/src/test/kotlin/com/sksamuel/scrimage/core/color/PixelToolsTest.kt
package com.sksamuel.scrimage.core.color

import com.sksamuel.scrimage.pixels.PixelTools
import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe
import java.awt.Point

class PixelToolsTest : FunSpec({

   val white: Int = 0xFFFFFFFF.toInt()
   val yellow: Int = 0xFFFFFF00.toInt()
   val pink: Int = 0xFFFFAFAF.toInt()

   test("non transparent alpha component") {
      white shouldBe java.awt.Color.WHITE.rgb
      PixelTools.alpha(white) shouldBe 255
   }

   test("transparent alpha component") {
      PixelTools.alpha(0xDD001122.toInt()) shouldBe 221
   }

   test("red component") {
      pink shouldBe java.awt.Color.PINK.rgb
      PixelTools.red(yellow) shouldBe 255
      PixelTools.red(pink) shouldBe 255
   }

   test("blue component") {
      pink shouldBe java.awt.Color.PINK.rgb
      PixelTools.blue(yellow) shouldBe 0
      PixelTools.blue(pink) shouldBe 175
   }

   test("green component") {
      yellow shouldBe java.awt.Color.YELLOW.rgb
      PixelTools.green(yellow) shouldBe 255
      PixelTools.green(pink) shouldBe 175
   }

   test("rgb combination") {
      PixelTools.rgb(255, 0, 255) shouldBe 0xFFFF00FF.toInt()
      PixelTools.rgb(85, 102, 119) shouldBe 0xFF556677.toInt()
   }

   test("coordinate to offset") {
      160 shouldBe PixelTools.coordsToOffset(10, 3, 50)
      10 shouldBe PixelTools.coordsToOffset(10, 0, 50)
      99 shouldBe PixelTools.coordsToOffset(49, 1, 50)
   }

   test("offset to coordinate") {
      PixelTools.offsetToPoint(0, 100) shouldBe Point(0, 0)
      PixelTools.offsetToPoint(100, 100) shouldBe Point(0, 1)
      PixelTools.offsetToPoint(99, 100) shouldBe Point(99, 0)
      PixelTools.offsetToPoint(199, 100) shouldBe Point(99, 1)
      PixelTools.offsetToPoint(101, 100) shouldBe Point(1, 1)
   }

})
